message/ipldbind/message.go

```diff
@@ -0,0 +1,83 @@
+package ipldbind
+
+import (
+	cid "github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/datamodel"
+
+	"github.com/ipfs/go-graphsync"
+)
+
+type MessagePartWithExtensions interface {
+	ExtensionNames() []graphsync.ExtensionName
+	Extension(name graphsync.ExtensionName) (datamodel.Node, bool)
+}
+
+type GraphSyncExtensions struct {
+	Keys   []string
```

Comment:
Something about bindnode I think, but it could be a hangover from trying to make this type replace the existing Extensions type that exposes a `ExtensionNames()`. It would be nice to remove this and just replace it with a `type GraphSyncExtensions map[string]datamodel.Node` (I haven't tried if that works, maybe it does?).

Reply:
Yes, this is bindnode; IPLD maps must be ordered, and Go maps aren't, so this is the best compromise I came up with.

Once Go 1.18 lands, I'll add a generic type "helper" to bindnode, as per https://github.com/ipld/go-ipld-prime/issues/340. Then, you'll be able to just do `bindnode.OrderedMap[string]datamodel.Node`.

One alternative would be if you said that you don't care about the order here, so a plain old Go map should be enough. And you're probably right. I've filed https://github.com/ipld/go-ipld-prime/issues/343 :)

Reply:
That said, I would say keep the current struct for now, because I don't think it's causing issues and it might take me a week or two to get to implementing support for plain maps.

Reply:
map key ordering, yeah that makes sense; don't prioritise changing this, we can stick with this